### PR TITLE
[Ray] Basic shuffle support for Ray backend

### PR DESCRIPTION
**Pull request by fyrestone**
## What do these changes do?

Most of the shuffle operands can provide the output data keys before execute, so it's possible to construct the Ray DAG for the shuffle operands.

This PR is the basic shuffle support for Ray backend, it based on the current shuffle operands and graph optimization. The push based shuffle will be the next gen: https://github.com/mars-project/mars/issues/2916

All the shuffle operands are list here, there are only two shuffle operands that generate the output data keys at runtime _(depend on the input data)_. This PR marks them as `To be implemented`. We can support these two operands in this inefficient way,

1. Generate the full output data key set before execute.
2. Construct the Ray DAG by `#1` keys.
3. Fill `None` value for the missing data keys before returning from `execute_subtask`.

<div class="lake-content" typography="classic">

Shuffle class | Can get the output data keys before execute
-- | --
mars.dataframe.align.DataFrameIndexAlign | True
mars.dataframe.base._duplicate.DuplicateOperand | True
mars.dataframe.base.drop_duplicates.DataFrameDropDuplicates | True
mars.dataframe.base.duplicated.DataFrameDuplicated | True
mars.dataframe.groupby.core.DataFrameGroupByOperand | True
mars.dataframe.groupby.sample.GroupBySample | True
mars.dataframe.merge.merge.DataFrameMergeAlign | True
mars.dataframe.sort.psrs.DataFramePSRSShuffle | True
mars.learn.ensemble._bagging.BaggingSample | True
mars.learn.utils.shuffle.LearnShuffle | True
mars.tensor.base.psrs.PSRSAlign | True
mars.tensor.base.psrs.PSRSShuffle | True
mars.tensor.base.unique.TensorUnique | True
mars.tensor.indexing.getitem.FancyIndexingConcat | True
mars.tensor.indexing.getitem.FancyIndexingDistribute | True
mars.tensor.indexing.setitem.TensorIndexSetValue | True
mars.tensor.random.permutation.TensorPermutation | True
mars.tensor.reshape.reshape.TensorReshape | False
mars.tensor.spatial.distance.pdist.TensorPdist | True
mars.tensor.spatial.distance.squareform.TensorSquareform | True
mars.tensor.statistics.bincount.TensorBinCount | False

</div>

## Related issue number

https://github.com/mars-project/mars/issues/2893

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/core/operand/shuffle.py [hunk 1]
```diff
@@ -40,6 +52,13 @@ def _new_chunks(self, inputs, kws=None, **kw):
 
         return super()._new_chunks(inputs, kws, **kw)
 
+    def get_output_data_keys(self) -> Union[ExactlyMapDataKeys, None]:
+        """
+        Get the output data keys before execute.
```
Comment (zhongchun) [inline mars/core/operand/shuffle.py:57]:
```suggestion
        Get the output data keys before execution.
```

File: mars/core/operand/shuffle.py [hunk 2]
```diff
@@ -22,6 +24,16 @@
 from .base import Operand, VirtualOperand, OperandStage
 
 
+class ExactlyMapDataKeys(list):
+    """
+    The type for the shuffle output data keys which
```
Comment (zhongchun) [inline mars/core/operand/shuffle.py:29]:
```suggestion
    The type for the shuffle output data keys is
```

File: mars/learn/utils/shuffle.py [hunk 3]
```diff
@@ -365,6 +366,28 @@ def execute_single(cls, ctx, op):
...
+            axes, reduce_sizes = self.axes, self.reduce_sizes
+            extra_axes, extra_reduce_sizes = [], []
+            for ax, reduce_size in zip(axes, reduce_sizes):
+                if reduce_size == 1:
+                    continue
+                extra_axes.append(ax)
+                extra_reduce_sizes.append(reduce_size)
+            axes, reduce_sizes = extra_axes, extra_reduce_sizes
+
+            keys = ExactlyMapDataKeys()
```
[the remark this answers is not in the document]
Comment (fyrestone) [reply mars/learn/utils/shuffle.py]: Fixed. Thanks.

File: mars/tensor/random/permutation.py [hunk 4]
```diff
@@ -154,6 +155,22 @@ def tile(cls, op):
             nsplits=nsplits,
         )
 
+    def get_output_data_keys(self):
+        if self.stage == OperandStage.map:
+            out_chunk = self.outputs[0]
+            reduce_size = self.reduce_size
+            keys = ExactlyMapDataKeys()
```
[the remark this answers is not in the document]
Comment (fyrestone) [reply mars/tensor/random/permutation.py]: Fixed.

File: mars/core/operand/shuffle.py (hunk 1 above)
Comment (fyrestone) [reply mars/core/operand/shuffle.py]: Fixed.

**Comment by qinxuye** [comment]
Need further discussions for this feature, thus convert to draft for now.

**Comment by fyrestone** [comment]
Close this PR because a new shuffle PR has been merged: https://github.com/mars-project/mars/pull/3040

**Pull request closed without merging**